File: nn/GeneticAlgorithm.cpp
#include "GeneticAlgorithm.h"



using namespace std;

char GeneticAlgorithm::wordLength=0;
typedef unsigned short ushort;


GeneticAlgorithm::GeneticAlgorithm(char seedSz, char chromosoneLn)
{
	seedSize = seedSz;
	chromosoneLength = chromosoneLn;

	weightFactor = .1;
	weightZero = 0;

	genePool = new short*[seedSize*seedSize];
	genePool[0] = NULL;
	
	chromosones = new Chromosone[seedSize*seedSize];
	
	initPool();


}




void GeneticAlgorithm::initPool() {

	// Clean up previous.
	deletePool();

	for(int i=0; i < seedSize * seedSize; i++){

		short * tmpArr =new short[chromosoneLength];

		for(int j = 0; j < chromosoneLength; j++){
		
			// +/- 2^wordLength. E.g., 10-bit = +/- (1024/2) 
			tmpArr[j] = getRandomClamped();	
			
		}
		genePool[i] = tmpArr;
		chromosones[i].genePoolPtr = tmpArr;
		// No need to delete [] tmpArr ptr **genePool[] has same addr;
	}

}

void GeneticAlgorithm::testInit(int cycles){

	for(int x=0;x< cycles;x++){

		initPool();

		for( int i=0; i<seedSize*seedSize; i++ ){		
		
			cerr << "Genome: \n";
			
			for( int j = 0; j<chromosoneLength; j++ ){
			
				cerr << (float)genePool[i][j] * weightFactor <<" ";
			}
			cerr << endl;
		}
		cerr << endl;

	}
}

void GeneticAlgorithm::deletePool(void){

	// Clean up.
	if(genePool[0]){
		for(int i=0; i < seedSize * seedSize ;i++){
			delete [] genePool[i];
		}
	}

}

short GeneticAlgorithm::getRandom(int min, int max){

	return rand()  % max + min/2; 

}

short GeneticAlgorithm::getRandomClamped(void){

	// +/- 2^wordLength/2
	return  getRandom(0, pow( 2.0,wordLength));// pow( 2.0, wordLength) / 2 - getRandom(0, pow( 2.0,wordLength));	


}

short GeneticAlgorithm::mutate(short& value) {

	// XOR random bit mask.

	// Get a random bit position 1~wordLength
	int mask = ( GeneticAlgorithm::getRandom(1, wordLength) );

	// Turn that bit on ... 2^(n-1) - 0,1,2,4,8...wordLength.
	mask = (int)floor(pow(2.0, (mask - 1)));     

	// XOR with value.
	value = value ^ mask;

	return 0;
}



bool GeneticAlgorithm::dumpPool(short start,short count){

	if (count == NULL) count =seedSize*seedSize; 

	if(genePool[0]){
		for(int i=start;i<start+count ;i++){		
			//		cerr << "Genome: \n";
			for(int j = 0; j<chromosoneLength; j++){
				cout << (float)genePool[i][j] ;
				if(j<chromosoneLength-1) cout <<",";
			}

			cout << endl;
		}
		cout << endl;
	}

	return 0;
}

int GeneticAlgorithm::epoch(){

	float crossoverRate = 1;
	float rnd = 0;
	int offset =0;
	int k = 0;
	mutationRate = .01	;
	
	// Push latent genes.
	// Apply roulette selection of fittess members
	
	
	// Step through parent gene pool.
	for (int i = 0; i < seedSize; i++) {

		// Inherit X from parent.
		short * x = genePool[i];

		// Process X against all other parent genomes
		for (int j=0 ; j< seedSize-1; j++) {

			int crossoverPoint = getRandom(0, wordLength*chromosoneLength);

			// Do not cross with self -n^(n-1) 
			//if (j == i) continue;

			rnd = getRandom(1,10);

			// Not tonight, I have a headache...
			if (rnd/10 > crossoverRate) {
			
				continue;
			}

			// Inherit Y from parent.
			short* y = genePool[j];

			// Apply crossover between two genomes.
			short *xy = new short[chromosoneLength]; 

			
	
		/////////////////
		
		// TODO ugly!
			int leftWords = (int)floor((double) (crossoverPoint/wordLength));
			
			
			for( k=0; k< leftWords ; k++ ){
			
				xy[k]=x[k];

					// Random mutaton.
				rnd = (float)getRandom(1,10);
		
				if( rnd/10 > mutationRate ){
			
					mutate(xy[k]);
					nMutations++;
			}
				
				
			}
		
			// Split point.
			if(leftWords){  
		//
		///////////////////////////	
							
				// Crossover X MSB.
				// 2^(n) -1 eg, 10-bit base = 1023 1111111111
				short base = pow( 2.0, wordLength ) -1;
				short msb = base - ( pow( 2.0, crossoverPoint % wordLength ) -1 );
				short lsb = base - msb;
				xy[k] = x[k] & msb;

				// Crossover Y LSB.
				xy[k] += y[k] & lsb;
				k++;
			}
	
			// Crossover remaining words from Y.
			for( ; k < chromosoneLength; k++ ){

				xy[k]=y[k];

				// Random mutaton.
				rnd = (float)getRandom(1,10)/10;
		
				if( rnd > mutationRate ){
			
						mutate(xy[k]);
						nMutations++;
				}
			
			}
		
			// Append new XY genome below parents. 
			for(int m =0; m<chromosoneLength;m++){

			genePool[ seedSize + offset][m] = xy[m];
			}
			
			offset++;
			//cerr << endl << crossoverPoint << endl;

			for(int m =0; m<chromosoneLength;m++){
			
				//cerr << x[m]<<" ";
			}
			
			//cerr << endl;
			
			for(int m =0; m<chromosoneLength;m++){
			
				//cerr << y[m]<<" ";
			}
			
			//cerr <<endl;
			
			for(int m =0; m<chromosoneLength;m++){
			
	//			cerr << genePool[ seedSize + (--offset)++][m] <<" ";
			}
		//cerr << endl;
			//cerr << endl;
			// Clean up.
			delete [] xy;
			
		}
		
		//cerr << endl; 
	
	}

	return 0;
}

GeneticAlgorithm::~GeneticAlgorithm(void)
{
	//delete stuff
}

File: js/GeneticAlgorithm.js
//  NeuralNetwork API Copyright 2011 Wayne Doucette, http://wayne-doucette.blogspot.com.
//  All rights reserved.
//  Simplified BSD License _Appllies.


function GeneticAlgorithm(chromosoneLength,wordLength) {

    // Genetic Algorithm
    this.wordLength =wordLength;
    this.mutationRate = .01;
    this.crossoverRate = 1;
    this.includeParents = true;
    this.weightFactor = 1 / 100;
    
    // Holds the population.
    this.pool = new Array();
    this.chromosoneLength = chromosoneLength;

}


GeneticAlgorithm.prototype.initPool = function (seed, poolSize) {

    // Initialize pool with random values.

    // wordLength - bits per geneome.
    // size - population of chromosones (returns ~size^2-size, depending on crossoverRate )
    // seed - sample chromosone
    this.chromosoneLength = seed.length;

    var genePool = new Array();

    for (i = 0; i < poolSize; i++) {

        var value = new Array();

        for (j = 0; j < seed.length; j++) {

            value.push(Math.floor(getRandom(0, 1) * Math.pow(2, this.wordLength)));
        }

        genePool.push(value);
    }

    this.pool = genePool;
}

GeneticAlgorithm.prototype.getTopChromosone = function (obj) {

    // calc fs
    this.evaluatePool(obj)
    return this.pool[0];

}

GeneticAlgorithm.prototype.epoch = function (genePool) {

    if (typeof genePool == 'undefined') genePool = this.pool.slice(0, 5);
    // Push latent genes.
    // genePool.push(latentGenome);

    var ngChromosones = new Array();

    // _Apply genetic crosssover 70% chance.
    // _Apply genetic mutation .1% chance.
    // _Apply roulette selection of fittess members
    // For each gene, crossover with other genes
    // with a possibility of mutation.


    // Step through parent gene pool.
    for (i = 0; i < genePool.length; i++) {
        var xChromosone = new Array();

        // Inherit xChromosone from parent.

        genePool[i].forEach(function (key, value) {

            xChromosone.push(value);
        });

        // Process against all other parent genomes
        for (j = 0; j < genePool.length; j++) {

            var yChromosone = new Array();
            var crossoverPoint = Math.round(getRandom(0, 1) * this.wordLength * this.chromosoneLength);

            // Do not cross with self
            //if (j == i) continue;


            // Not tonight, I have a headache...
            if (getRandom(0, 1) > this.crossoverRate) continue;

            // Inherit yChromosone from parent.
            genePool[j].forEach(function (key, value) {

                yChromosone.push(value);

            });


            // _Apply crossover between genomes creating a next-generation chromosone.

            var xy = new Array();
            var leftWords = Math.floor(crossoverPoint / this.wordLength);

            for (k = 0; k < leftWords; k++) {

                xy[k] = xChromosone[k];
                var rnd = getRandom(0, 1); // Math.floor(Math.random() /this.mutationRate) + 1;

                if (rnd > (this.mutationRate)) {

                    xy[k] = this.mutate(xy[k]);
                    this.gMutations++;

                }
            }

            // Split point.
            if (leftWords) {
                //
                ///////////////////////////	

                // Crossover X MSB.
                // 2^(n) -1 eg, 10-bit base = 1023 1111111111
                var base = Math.pow(2.0, this.wordLength) - 1;

                // known bug
                var msb = base - (Math.pow(2.0, this.wordLength - crossoverPoint % this.wordLength) - 1);
                var lsb = base - msb;

                xy[k] = (xChromosone[k] & msb);

                // Crossover Y LSB.
                xy[k] += (yChromosone[k] & lsb);

                // Random mutaton.
                var rnd = getRandom(0, 1);

                if (rnd > (this.mutationRate)) {

                    xy[k] = this.mutate(xy[k]);
                    this.gMutations++;
                }



                k++;
            }

            // Crossover remaining words from Y.
            for (; k < this.chromosoneLength; k++) {

                xy[k] = yChromosone[k];

                // Random mutaton.
                var rnd = getRandom(0, 1);

                if (rnd > (this.mutationRate)) {

                    xy[k] = this.mutate(xy[k]);
                    this.gMutations++;
                }

            }

            //status(gMutations);
            ngChromosones[ngChromosones.length] = new Array();

            xy.forEach(function (key, value) {

                ngChromosones[ngChromosones.length - 1].push(value);

            });

        }


    }

    // Include parents - Parent pool survives WITHOUT evolving.

    // TODO: hardcoded seedSize (5):
    if (this.includeParents) ngChromosones = ngChromosones.concat(genePool.slice(0, 5));

    // TODO: ? Persist existing fitness scores. 
    for (var i = 0; i < 5; i++) {
        //ngChromosones[i].fs = genePool[i].fs;
    }
    this.pool = ngChromosones;

}

GeneticAlgorithm.prototype.getWeights = function () {


    var result = new Array();

    this.pool.forEach(function (key, value) {

        // Convert (10-bit) number into +/- decimal.

        // TODO Decimal place adapts to word length. 
        result.push(((value - (Math.pow(2, wordLength) * .5)) * weightFactor));

        //return ((this.weights[index] - (Math.pow(2, wordLength) / 2))).toFixed(2) //this.weights[index];

    });

    return result;

}

GeneticAlgorithm.prototype.getIndexToBinaryString = function (value) {


    var binary = new Array();
    var out = value.toString(2);
    while (out.length < this.wordLength) out = "0" + out;
    binary.push(out);

    return binary;

}

GeneticAlgorithm.prototype.getBinaryString = function (index) {


    var binary = new Array();
    this.pool[index].forEach(function (key, value) {

        var out = parseFloat(value).toString(2);
        while (out.length < this.wordLength) out = "0" + out;
        binary.push(out);
    });

    return binary;

}

GeneticAlgorithm.prototype.evaluatePool = function (obj) {

    var pool = this.pool;

    // Each chromosone.

    pool.forEach(function (pKey, value, obj) {
     // Reset fitness.
        pool[pKey].fs = 0;

        // Each training set.
        trainingSets.forEach(function (tsKey, value, obj) {

            inputs = trainingSets[tsKey].inputs;
            targets = trainingSets[tsKey].outputs;

            var results = obj.update(inputs, pool[pKey]);

            // Evaluate fitness and set new score.
           results.forEach(function (index, value) {

                var delta = Math.abs(targets[index] - results[index]);

                // LMS 
                delta *= delta;

                pool[pKey].fs += delta;

            });

        }, obj);

    }, obj);


    // Order pool by lowest (best) fitness score.

    pool.sort(function (a, b) {
        return (a.fs) - (b.fs);
    });

    return;

}



GeneticAlgorithm.prototype.mutate = function(value) {

    // Returns value with bitwise mutation -flips a random bit. 

    var mask;

    // 1 ~ wordLength
    mask = Math.round(Math.random() * Math.ceil(Math.log(Math.pow(2, this.wordLength)) / Math.log(2) - 1));

    // 2^(n-1) - 0,1,2,4,8...
    mask = Math.floor(Math.pow(2, (mask - 1)));

    return (value ^ mask);

}
//END GA

